webapp/sql/create_pois_table.sql

```diff
@@ -0,0 +1,70 @@
+DROP TABLE IF EXISTS pois;
+
+CREATE TABLE pois (
+	id SERIAL PRIMARY KEY,
+	geometry geometry,
+);
+
+-- Note that building filters some pois such as sports_centre because sports 
+-- centre is not a building
+CREATE OR REPLACE VIEW building AS
+	SELECT *
+	FROM planet_osm_polygon AS building
+	WHERE (
+		building IS NOT NULL
+		AND access IS DISTINCT FROM 'private'
+	);
+		
+INSERT INTO pois(geometry, osm_id, is_node) (
+	SELECT way, pois_subquery.osm_id, pois_subquery.is_node
+	FROM (
+		SELECT
+			CASE 
+				WHEN building.way IS NULL THEN point.way
+				ELSE building.way
+			END, point.amenity, point.shop, point.leisure, 
+			point.landuse, point.access
+		FROM building RIGHT JOIN planet_osm_point AS point 
+		ON ST_Within(point.way, building.way)
+		
+		UNION ALL
```

Comment:
personally I would split the queries into two queries

Reply:
this is just pseudo code, but a small proposal:

```sql
DROP TABLE IF EXISTS pois;

CREATE TABLE pois (
	id SERIAL PRIMARY KEY,
	geometry geometry,
);

CREATE VIEW poi_nodes (
  SELECT * FROM planet_osm_nodes WHERE **tag_condition**;
)

-- Insert all POI polygons from planet_osm_plygon into pois table
INSERT INTO pois(geometry, osm_id, is_node) (
		SELECT way AS geometry, amenity, shop, leisure, landuse, "access"
		FROM planet_osm_polygon
    WHERE (**tags_condition**)
          AND "access" IS DISTINCT FROM 'private'
    )
);

-- eventually raise notice?

-- Insert polygons, which have a POI node on it, into pois table
-- the polygon is either a building or a buffer
INSERT INTO pois(geometry, osm_id, is_node) (
  SELECT
			CASE 
				WHEN building.way IS NULL THEN ST_Buffer(point.way, **???**)
				ELSE building.way
		FROM building RIGHT JOIN planet_osm_point AS point 
		ON ST_Within(point.way, building.way)
)

-- eventually raise notice?

CREATE INDEX pois_geometry ON pois USING gist(geometry);

```

Reply:
i think it is important that the query can be understood fast. So I would add some comments which clarify what is going on (as i did the in the pseudo code, but they can be more accurate and comprehensive).

Reply:
what the `**tags_condition**` should be, i don't know myself...  the condition could be passed from outside... but this could be over complicated... maybe going back to the duplication would be the clearest? :worried:

Reply:
@sfkeller What do you as our (Postgre)SQL expert think?